xds/src/main/java/io/grpc/xds/ClientXdsClient.java

```diff
@@ -361,15 +365,143 @@ static FilterChain parseFilterChain(
     if (name.isEmpty()) {
       name = UUID.randomUUID().toString();
     }
+    FilterChainMatch filterChainMatch = parseFilterChainMatch(proto.getFilterChainMatch());
+    checkForUniqueness(uniqueSet, filterChainMatch);
     return new FilterChain(
         name,
-        parseFilterChainMatch(proto.getFilterChainMatch()),
+        filterChainMatch,
         httpConnectionManager,
         downstreamTlsContext,
         tlsContextManager
     );
   }
 
+  private static void checkForUniqueness(Set<FilterChainMatch> uniqueSet,
+      FilterChainMatch filterChainMatch) throws ResourceInvalidException {
+    if (uniqueSet != null) {
+      Set<FilterChainMatch> crossProduct = getCrossProduct(filterChainMatch);
+      for (FilterChainMatch cur : crossProduct) {
+        if (uniqueSet.contains(cur)) {
+          throw new ResourceInvalidException("Found duplicate matcher: " + cur);
+        }
+      }
+      uniqueSet.addAll(crossProduct);
+    }
+  }
+
+  private static Set<FilterChainMatch> getCrossProduct(FilterChainMatch filterChainMatch) {
+    // repeating fields to process:
+    // prefixRanges, applicationProtocols, sourcePrefixRanges, sourcePorts, serverNames
+    Set<FilterChainMatch> expandedSet = expandOnPrefixRange(filterChainMatch);
+    expandedSet = expandOnApplicationProtocols(expandedSet);
+    expandedSet = expandOnSourcePrefixRange(expandedSet);
+    expandedSet = expandOnSourcePorts(expandedSet);
+    return expandOnServerNames(expandedSet);
+  }
+
+  private static Set<FilterChainMatch> expandOnPrefixRange(FilterChainMatch filterChainMatch) {
+    if (filterChainMatch.getPrefixRanges().isEmpty()) {
+      return Collections.singleton(filterChainMatch);
+    }
+    HashSet<FilterChainMatch> expandedSet = new HashSet<>(
+        filterChainMatch.getPrefixRanges().size());
+    for (EnvoyServerProtoData.CidrRange cidrRange : filterChainMatch.getPrefixRanges()) {
+      expandedSet.add(new FilterChainMatch(filterChainMatch.getDestinationPort(),
+          Arrays.asList(cidrRange),
+          ImmutableList.copyOf(filterChainMatch.getApplicationProtocols()),
```

Comment:
Do you really need make a copy? Would `Collections.unmodifiableList()` work?

Reply:
Good point. Don't need the copy. Let me fix

Reply:
Done